### PR TITLE
Add TrainEngine abstraction to TorchTrainStep, add FairScale integration, improve transformers integration

### LINKED ISSUE #31: DeepSpeed integration

### PR DESCRIPTION
**Pull request by epwalsh**
## Overview

This started out as attempt to integrate DeepSpeed, but I got frustrated and switched to FairScale for now. I'll follow up with another PR to add the DeepSpeed integration when I'm satisfied with it.

Changes proposed in this pull request:
- Adds a `TrainEngine` class to torch integration which defines and drives the training strategy, similar to PyTorch Lightning's `Strategy` abstraction.
- Adds a FairScale integration with a `FairScaleTrainEngine` implementation that leverages FairScale's `FullyShardedDataParallel` (FSDP) for ZeRO-3 optimizations during distributed training.
- Adds a lot of useful stuff to the transformers integration (registered models, tokenizers, data collators, etc.).
- Many improvements to documentation of our torch-related integrations.
- A couple bug fixes in `FromParams`.

## Empirical Results of FairScale Experiments

I'm pleasantly surprised with the performance of FairScale so far. Here are some results for training a causal language model on sequences of length 1024 using various model sizes, machines, and configurations.

Note that experiments within each round achieved a similar loss.

### Round 1

In this round I wanted to see how much memory savings we could get with a small model. You can also see these results on [Weights & Biases](https://wandb.ai/allennlp/tango-fairscale-benchmarks/workspace), or run them yourself with:

```bash
tango run integration_tests/fairscale_benchmarks/config.jsonnet -i examples/train_lm/tokenize_step.py
```

**Model:** [gpt2](https://huggingface.co/gpt2)
**Environment:** [allennlp-cirrascale-01](https://beaker.org/cl/ai2/allennlp-cirrascale/details) machine using 4x80GiB A100 GPUs, running in Beaker interactive

| ID | FSDP | AMP | Activation Checkpointing | CPU Offloading | Max Mem Used Per GPU | Seconds Per Batch |
| -- | :---: | :-----: | :------------------------: | :--------------: | ------------------------: | --------------------: |
| gpt2-A | – | – | – | – | 31920 MiB | 1.19 |
| gpt2-B | ✓ | – | – | – | **30171** MiB | 1.21 |
| gpt2-C | ✓ | ✓ | – | – | **24198** MiB | **0.48** |
| gpt2-D | ✓ | ✓ | ✓ | – | **5316** MiB | 1.55 | 
| gpt2-E | ✓ | ✓ | ✓ | ✓ | **4837** MiB | 3.51 |
| gpt2-F | – | ✓ | – | – | **26273** MiB | **0.49** |
| gpt2-G | – | ✓ | ✓ | – | **7231** MiB | 1.91 |

**Take-aways:** With this small model size, I believe most of the memory usage was due to activations (the batch size was fairly large), which explains why FSDP by itself didn't have much of impact on memory (compare "A" and "B") but activation checkpointing had a huge impact (compare "C" and "D"). AMP also had a significant impact on memory as well as compute time, both with and without FSDP.

### Round 2

This time I wanted to push the limits and try training a much larger model. Most of the configurations below resulting in CUDA OOM errors (indicated by N/A). You can also reproduce these with the same command:

```bash
tango run integration_tests/fairscale_benchmarks/config.jsonnet -i examples/train_lm/tokenize_step.py
```

But first you'll need to tweak the configuration file. The comments in `config.jsonnet` should give you all of the information you need to change the model to GPT-J 6B.

**Model:** [EleutherAI/gpt-j-6B](https://huggingface.co/EleutherAI/gpt-j-6B)
**Environment:** [allennlp-cirrascale-01](https://beaker.org/cl/ai2/allennlp-cirrascale/details) machine using 4x80GiB A100 GPUs, running in Beaker interactive

| ID | FSDP | AMP | Activation Checkpointing | CPU Offloading | Max Mem Used Per GPU | Seconds Per Batch |
| -- | :---: | :-----: | :------------------------: | :--------------: | ------------------------: | --------------------: |
| gptj-A | – | – | – | – | N/A | N/A |
| gptj-B | ✓ | – | – | – | N/A | N/A |
| gptj-C | ✓ | ✓ | – | – | N/A | N/A |
| gptj-D | ✓ | ✓ | ✓ | – | **34403** MiB | **7.06** | 
| gptj-E | ✓ | ✓ | ✓ | ✓ | **13023** MiB | **17.75** |
| gptj-F | – | ✓ | – | – | N/A | N/A |
| gptj-G | – | ✓ | ✓ | – | N/A | N/A |
| gptj-H | ✓ | – | ✓ | – | **36919** MiB | **10.71** | 

**Take-aways:** FSDP with Activation Checkpointing was required to train this thing (runs "D", "E", and "H"). Comparing "D" and "H", it's a no-brainer to add in AMP since that leads to some memory savings and faster training time. CPU offloading gives a huge boost in memory savings, but is 2.5 times slower (compare "E" and "D").

### CODE REVIEW

File: tests/integrations/torch/train_test.py [hunk 1]
```diff
@@ -48,6 +57,7 @@ def test_basic_train_with_streaming_data(self):
         assert (result_dir / "train" / "data.pt").is_file()
 
     def test_train_distributed(self):
+        initialize_logging()
```
Comment (epwalsh) [inline tests/integrations/torch/train_test.py:60]: TODO: remove once https://github.com/allenai/tango/pull/177 merges
```suggestion
```

**Comment by dirkgr** [comment]
> CPU offloading gives a huge boost in memory savings, but is 2.5 times slower (compare "E" and "D").

What was the batch size? Can you recover the speed because you can get away with bigger batches?

**Comment by epwalsh** [comment]
> What was the batch size? Can you recover the speed because you can get away with bigger batches?

Batch size was 32. In this case I didn't want a higher batch size since it would lead to over-fitting, so I don't have an answer for that yet.

File: tango/integrations/torch/train_callback.py [hunk 2]
```diff
@@ -21,14 +21,18 @@ class TrainCallback(Registrable):
...
+
     .. seealso::
         See :class:`~tango.integrations.wandb.WandbTrainCallback` for an example
         implementation.
 
     Attributes
     ----------
     train_config : :class:`TrainConfig`
-    model : :class:`Model`
+    training_engine : :class:`TrainingEngine`
```
Comment (dirkgr) [inline tango/integrations/torch/train_callback.py:35]: Is that change necessary? If `TrainCallback` becomes a major integration point, it will be easier to understand how to write one if you get a `Model` instead of a `TrainingEngine`. But that's only true if those callbacks don't need access to the engine.

File: tango/integrations/torch/training_engine.py [hunk 3]
```diff
@@ -0,0 +1,318 @@
...
+    Parameters
+    ----------
+    amp : :class:`bool`, optional
+        Use automatic mixed precision. Default is ``False``.
+    max_grad_norm : :class:`float`, optional
+        If set, gradients will be clipped to have this max norm. Default is ``None``.
+    amp_use_bfloat16 : ``Optional[bool]``
+        Set to ``True`` to force using the ``bfloat16`` datatype in mixed precision training.
+        Only applicable when ``amp=True``. If not specified, the default behavior will be
+        to use ``bfloat16`` when training with AMP on CPU, otherwise not.
```
Comment (dirkgr) [inline tango/integrations/torch/training_engine.py:155]: also `lr_scheduler`

File: tango/integrations/transformers/__init__.py [hunk 4]
```diff
@@ -5,6 +5,148 @@
...
+            for name in sorted(LRScheduler.list_available()):
+                if name.startswith("transformers::"):
+                    print(name)
+
+        .. testoutput::
+            :options: +ELLIPSIS
+
+            transformers::constant
+            transformers::constant_with_warmup
+            transformers::cosine
```
Comment (dirkgr) [inline tango/integrations/transformers/__init__.py:103]: There should be a lot more: https://github.com/huggingface/transformers/blob/v4.15.0/src/transformers/optimization.py#L272

**Review by dirkgr — CHANGES_REQUESTED** [review]
I feel like `TrainingEngine` is getting close to the point at which we could just have different trainers.

File: .github/workflows/main.yml [hunk 5]
```diff
@@ -69,19 +69,29 @@ jobs:
               pytest -v --color=yes --doctest-modules tango/integrations/datasets tests/integrations/datasets
 
           - name: PyTorch integration
-            extras: dev,torch
+            extras: dev,all
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply .github/workflows/main.yml]: Yea, you're right. I changed it back: 4818902

File: examples/train_lm/tokenize_step.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            tokenize_function,
+            batched=True,
+            num_proc=num_workers,
+            remove_columns=["text"],
+            desc="Tokenizing dataset",
+        )
+
+        def group_texts(examples):
+            # Concatenate all texts.
+            concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}  # type: ignore
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply examples/train_lm/tokenize_step.py:36]: The only keys are `input_ids` and `attention_mask`.

File: tango/integrations/fairscale/module_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+import torch
+import torch.nn as nn
+from fairscale.nn.checkpoint import checkpoint_wrapper
+
+from tango.integrations.torch import Model
+
+from .fsdp_config import FSDPConfig
+
+
+@Model.register("fairscale::with_wrapped_modules")  # type: ignore[arg-type]
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/module_wrapper.py:13]: You need to use this wrapper explicitly on all models. See https://github.com/allenai/tango/blob/f7facd430479756a6d183bcb58746559879ff153/examples/train_lm/config.jsonnet#L91-L101

File: tango/integrations/fairscale/training_engine.py [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+            amp_use_bfloat16=amp_use_bfloat16,
+        )
+        if amp:
+            self.grad_scaler = ShardedGradScaler()
+
+    def _construct_model(self, model: Lazy[Model]) -> Model:
+        model: Model = model.construct()
+        if not self.fsdp_config.move_params_to_cpu:
+            model.to(self.train_config.worker_local_default_device)
+        return FSDP(model, **self.fsdp_config.as_kwargs())
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/training_engine.py:129]: > This does not look like it supports efficient loading of models into GPU memory?

Why not?

[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/training_engine.py:129]: If you use `with_wrapped_modules()` on the inner modules of your model, it will be efficient. Maybe I forgot to write that in the docs explicitly.

[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/training_engine.py:129]: e83f972

File: tango/integrations/torch/train.py [hunk 9]
```diff
@@ -462,66 +427,52 @@ def save_state(step: int):
...
-            temp_state_file.close()
-            os.replace(temp_state_file.name, config.state_path_for_step(step))
-
-            # Link to most recent state path.
-            if config.state_path.is_file():
+
+        # Link to most recent state path.
+        # NOTE: While hard linking would be preferable to creating symlinks, some train engines
+        # require a whole directory to save their state instead of a single file, which
+        # means state_path_for_step will be a directory, so a hard link won't work.
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/torch/train.py:448]: The latest and best checkpoint directory will never be removed: https://github.com/allenai/tango/blob/e83f9721c4c9363409edba587e602222a28cf5ae/tango/integrations/torch/train.py#L462-L465

File: tango/integrations/transformers/__init__.py (hunk 4 above)
Comment (epwalsh) [reply tango/integrations/transformers/__init__.py:103]: There ARE a lot more. These are just the first 3, alphabetically. Notice the ellipsis on the next line.

File: tango/integrations/torch/train_callback.py (hunk 2 above)
Comment (epwalsh) [reply tango/integrations/torch/train_callback.py:36]: Well, a `TrainingEngine` is responsible for constructing the `Model` (since it might have its own, more memory efficient way of doing so), so I think it makes sense for the `TrainingEngine` to own the model. That said, I see your point. It's important that each `TrainCallback`s can access the model. And they can, since part of the `TrainingEngine` API is that the model is accessible through [`TrainingEngine.model`](https://github.com/allenai/tango/blob/ee179adff1927c1cc0aaae84db9b23b2759b6a06/tango/integrations/torch/training_engine.py#L26). But this should be more obvious. Here's my solution to that: 12294d7

File: tango/integrations/torch/train.py [hunk 10]
```diff
@@ -156,10 +154,6 @@ def run(  # type: ignore[override]
             Save a checkpoint every this many steps.
         validate_every : :class:`int`, optional
             Run the validation loop every this many steps.
-        amp : :class:`bool`, optional
-            Use automatic mixed precision. Default is ``False``.
-        max_grad_norm : :class:`float`, optional
-            If set, gradients will be clipped to have this max norm. Default is ``None``.
```
[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/torch/train.py:162]: `Optimizer` and `LRScheduler` are both engine settings already, but they just happen to specified at the `Step` level. After reviewing this just now, there's no good reason not to just specify them at the engine level instead. Fixed in 10567b3.

I can see why `grad_accum` could make sense as a `TrainingEngine` parameter, but it has actually has more to do with data loading, which happens outside of the `TrainingEngine`.

File: tango/integrations/torch/training_engine.py (hunk 3 above)
Comment (epwalsh) [reply tango/integrations/torch/training_engine.py:155]: https://github.com/allenai/tango/commit/10567b387e9d074e3854df5f72d37c06fe95cea4

File: tango/integrations/transformers/__init__.py (hunk 4 above)
Comment (dirkgr) [reply tango/integrations/transformers/__init__.py:103]: 👍🏻

File: tango/integrations/torch/train_callback.py (hunk 2 above)
Comment (dirkgr) [reply tango/integrations/torch/train_callback.py:36]: Can you add a note to the docs for `TrainCallback` that this is how you get to the model? I can imagine an impatient coder looking at the docs for `TrainCallback` and not finding it. I can imagine that because I have been that coder.

File: tango/integrations/torch/train.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/torch/train.py:162]: So `grad_accum` gets passed to the `TrainingEngine` as one of the `Lazy` parameters?

File: tango/integrations/torch/train.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/torch/train.py:431]: :+1:

File: tango/integrations/fairscale/module_wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/fairscale/module_wrapper.py:13]: In some places, huggingface has `if deepspeed:` lines. You could imagine that they have `if fairscale: self.something = fairscale_wrapper(self.something)`. They do that kind of thing sometimes. But I don't know if they do it this time.

File: examples/train_lm/tokenize_step.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply examples/train_lm/tokenize_step.py:36]: That step can run on all kinds of datasets though. What if I want to train an LM on SQuAD text?

File: tango/integrations/fairscale/training_engine.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/fairscale/training_engine.py:129]: I just see the naked `model.construct()`. At that point, no code that's fairscale specific has run yet. So if `model.construct()` takes 500 GB of memory, we'll need 500 GB of memory on that line.

[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/training_engine.py:129]: See my previous two comments

File: tango/integrations/torch/train_callback.py (hunk 2 above)
Comment (epwalsh) [reply tango/integrations/torch/train_callback.py:39]: Done: 62a8ad9

File: tango/integrations/fairscale/module_wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (epwalsh) [reply tango/integrations/fairscale/module_wrapper.py:13]: I can only find them doing this at the Model level, i.e. the outermost level: https://github.com/huggingface/transformers/blob/ba3f9a71a1d0ac30b2f469f426f350fe6f374de7/src/transformers/trainer.py#L1062

File: examples/train_lm/tokenize_step.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (epwalsh) [reply examples/train_lm/tokenize_step.py:37]: 68be712

File: tango/integrations/torch/train.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/torch/train.py:162]: I think DeepSpeed needs it, but I'm fine adding it later.

File: tango/integrations/fairscale/training_engine.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/fairscale/training_engine.py:129]: I did not get this before. But I think now I do. Since the output of `model.construct()` is already wrapped, it's not true that nothing fairscale specific has run. The module wrappers have already been applied, so it's all good.

File: tango/integrations/fairscale/module_wrapper.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (dirkgr) [reply tango/integrations/fairscale/module_wrapper.py:13]: So there are two ways of doing it, relying on HF's use of the auto-wrapper, and our `with_wrapped_modules`. I'm fine with that situation, I just want to make sure that not something stupid happens where HF's automation will try to wrap the model again or something.

**Review by dirkgr — APPROVED** [review]
